src/target/riscv/riscv-013.c

```diff
@@ -2811,6 +2818,324 @@ void riscv013_fill_dmi_nop_u64(struct target *target, char *buf)
+			else
+				LOG_ERROR("System Bus Access Test 5: SBCS sbaccess error test FAILED, unable to clear to 0.");
+		} else {
+			LOG_ERROR("System Bus Access Test 5: SBCS sbaccess error test FAILED, unable to set error code.");
+		}
+	}
+
+	/* Test 6: Write to misaligned address */
+	sbcs = set_field(sbcs_orig, DMI_SBCS_SBACCESS, 1);
+
+	write_memory_sba_simple(target, legal_address+1, test_patterns, 1, sbcs);
+
+	dmi_read(target, &rd_val, DMI_SBCS);
+	if (get_field(rd_val, DMI_SBCS_SBERROR) == 3) {
+		sbcs = set_field(sbcs_orig, DMI_SBCS_SBERROR, 1);
+		dmi_write(target, DMI_SBCS, sbcs);
+		dmi_read(target, &rd_val, DMI_SBCS);
+		if (get_field(rd_val, DMI_SBCS_SBERROR) == 0)
+			LOG_INFO("System Bus Access Test 6: SBCS address alignment error test PASSED");
+		else
+			LOG_ERROR("System Bus Access Test 6: SBCS address alignment error test FAILED, unable to clear to 0.");
+	} else {
+		LOG_ERROR("System Bus Access Test 6: SBCS address alignment error test FAILED, unable to set error code.");
+	}
+
+	/* Test 7: Set sbbusyerror, only run this case in simulation as it is likely
+	 * impossible to hit otherwise */
+	if (run_sbbusyerror_test) {
+		sbcs = set_field(sbcs_orig, DMI_SBCS_SBREADONADDR, 1);
+		dmi_write(target, DMI_SBCS, sbcs);
+
+		for (int i = 0; i < 16; i++)
+			dmi_write(target, DMI_SBDATA0, 0xdeadbeef);
+
+		for (int i = 0; i < 16; i++)
+			dmi_write(target, DMI_SBADDRESS0, legal_address);
+
+		dmi_read(target, &rd_val, DMI_SBCS);
+		if (get_field(rd_val, DMI_SBCS_SBBUSYERROR)) {
+			sbcs = set_field(sbcs_orig, DMI_SBCS_SBBUSYERROR, 1);
+			dmi_write(target, DMI_SBCS, sbcs);
+			dmi_read(target, &rd_val, DMI_SBCS);
+			if (get_field(rd_val, DMI_SBCS_SBBUSYERROR) == 0)
+				LOG_INFO("System Bus Access Test 7: SBCS sbbusyerror test PASSED.");
+			else
+				LOG_ERROR("System Bus Access Test 7: SBCS sbbusyerror test FAILED, unable to clear to 0.");
+		} else {
+			LOG_ERROR("System Bus Access Test 7: SBCS sbbusyerror test FAILED, unable to set error code.");
+		}
+	}
+
+	return ERROR_OK;
+
+}
+
+void write_memory_sba_simple(struct target *target, target_addr_t addr,
+		uint32_t *write_data, uint32_t write_size, uint32_t sbcs)
+{
+	RISCV013_INFO(info);
+
+	uint32_t rd_sbcs;
+	uint32_t masked_addr;
+
+	uint32_t sba_size = get_field(info->sbcs, DMI_SBCS_SBASIZE);
+
+	read_sbcs_nonbusy(target, &rd_sbcs);
+
+	uint32_t sbcs_no_readonaddr = set_field(sbcs, DMI_SBCS_SBREADONADDR, 0);
+	dmi_write(target, DMI_SBCS, sbcs_no_readonaddr);
+
+	for (uint32_t i = 0; i < sba_size/32; i++) {
+		masked_addr = (addr >> 32*i) & 0xffffffff;
+
+		if (i != 3)
+			dmi_write(target, DMI_SBADDRESS0+i, masked_addr);
+		else
+			dmi_write(target, DMI_SBADDRESS3, masked_addr);
+	}
+
+	/* Write SBDATA registers starting with highest address, since write to
+	 * SBDATA0 triggers write */
+	for (int i = write_size-1; i >= 0; i--)
+		dmi_write(target, DMI_SBDATA0+i, write_data[i]);
+}
+
+uint32_t *read_memory_sba_simple(struct target *target, target_addr_t addr,
```

Comment:
i think you should take the `read_data` in as a buffer here rather than mallocing it in this function. That seems likely to be overkill, as you could just have one buffer in the calling function that you use for everything